src/pybind/mgr/dashboard/services/service.py

```diff
@@ -54,42 +55,83 @@ def wait_for_refresh(orch, service_name, last_refreshed):
+                raise DashboardException(
+                    code='daemon_list_timeout',
+                    msg="Timeout reached while waiting for daemon list for service %s." % service_name
+                )
+            time.sleep(1)
+            continue  # Retry getting daemon list
+
         all_running = True
 
         for daemon in daemons:
+            logger.info("Checking daemon %s state", daemon['daemon_id'])
             daemon_state = daemon['status_desc']
-            logger.debug("Daemon %s state: %s", daemon['daemon_id'], daemon_state)
+            logger.info("Daemon %s state: %s", daemon['daemon_id'], daemon_state)
 
             if daemon_state in ('unknown', 'error', 'stopped'):
-                logger.error("Failed to restart daemon %s for service %s. State is %s", daemon['daemon_id'], service_name, daemon_state)  # noqa E501  # pylint: disable=line-too-long
+                logger.error("Failed to restart daemon %s for service %s. State is %s", daemon['daemon_id'], service_name, daemon_state)
                 raise DashboardException(
                     code='daemon_restart_failed',
-                    msg="Failed to restart the daemon %s. Daemon state is %s." % (service_name, daemon_state)  # noqa E501  # pylint: disable=line-too-long
+                    msg="Failed to restart the daemon %s. Daemon state is %s." % (daemon['daemon_id'], daemon_state)
                 )
-            if daemon_state != 'running':
+
+            if daemon_state == 'starting':
+                all_running = False
+
+            elif daemon_state != 'running':
                 all_running = False
 
         if all_running:
             logger.info("All daemons for service %s are running", service_name)
             return True
 
-        if time.time() - start_time > 10:
+        if time.time() - start_time > timeout:
             logger.error("Timeout reached while waiting for daemon %s to start", service_name)
             raise DashboardException(
                 code='daemon_restart_timeout',
                 msg="Timeout reached while waiting for daemon %s to start." % service_name
             )
-        return False
+
+        time.sleep(1)  # Adding a short delay before retrying
 
 
 class RgwServiceManager:
+    def create_rgw_instance(self, service_spec):
+        # Create RGW instance.
+        logger.info("Creating RGW instance")
+        orch = OrchClient.instance()
+        result = orch.services.apply(service_spec)
+        return result
+    
+    def find_available_port(self, starting_port=80):
+        orch = OrchClient.instance()
+        daemons = [d.to_dict() for d in orch.services.list_daemons(daemon_type='rgw')]
+        used_ports = set()
+        for daemon in daemons:
+            ports = daemon.get('ports', [])
+            if ports:
+                used_ports.update(ports)
+        port = starting_port
+        while port in used_ports:
+            port += 1
+        return port
```

Comment:
@epuertat I agree that this is something Dashboard code/logic shouldn't do. Besides, this could could fail for other reason for instance if a port is not listed it doesn't mean that that port is not in use in the destination node by some other service (other than rgw or directly not controlled by cephadm!). I think it's a good thing to discuss in cephadm weekly and see what's the best approach to provide that support from cephadm. I'll add it in tomorrow's weekly agenda :+1: